use _retryWithInterval where possible

RaananW opened the pull request:
This is an example of how _retryWithInterval can be used instead of setTimout of the calling function.

There are a few other places in the framework that could benefit from that, but looking for feedback whether or not it makes sense to do the work.

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

bjsplat commented:
Snapshot stored with reference name:
refs/pull/15988/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15988/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15988/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15988/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15988/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15988/merge#BCU1XR#0

sebavan commented:
I think we should move the function in its own module but I like it :-) 

The timeout is the nice part but I am not sure what the value should be.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl2playwright/index.html

bjsplat commented:
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgpuplaywright/index.html

RaananW commented:
> I think we should move the function in its own module but I like it :-)
> 
> The timeout is the nice part but I am not sure what the value should be.

Moved, and added a few other places where it might make sense to use it.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

sebavan commented:
Let s see what others think but I quite like the factorization and the additional timeout  (still a question of the default like scene isready and shader compilation could take way longer) cause at least we wont leave leftover in the scene.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15988/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

RaananW commented on packages/dev/core/src/Misc/timingTools.ts:
```diff
@@ -17,3 +17,48 @@ export class TimingTools {
+            maxTimeout -= step;
+            if (maxTimeout < 0) {
+                clearInterval(int);
+                onError?.(new Error("Operation timed out after maximum retries"));
```
there was a request to add here an extra optional string

sebavan reviewed (changes requested):
Let s check together the default timeout but shaders for instance should be up to 30 seconds

RaananW commented:
> Let s check together the default timeout but shaders for instance should be up to 30 seconds

This was done in a different PR, also updated here.

sebavan reviewed (commented):
I guess the default compute values should also be 30s ?

RaananW commented:
> I guess the default compute values should also be 30s ?

Already done. what's different there is the timeout (interval call length), and not the max timeout

Popov72 commented on packages/dev/core/src/Misc/timingTools.ts:
```diff
@@ -17,3 +17,49 @@ export class TimingTools {
+            maxTimeout -= step;
+            if (maxTimeout < 0) {
+                clearInterval(int);
+                onError?.(new Error("Operation timed out after maximum retries. " + additionalStringOnTimeout));
```
```suggestion
                onError?.(new Error("Operation timed out after maximum retries. " + (additionalStringOnTimeout || "")));
```
